pkg/controller/envoy/ads_event_test.go

```diff
@@ -0,0 +1,637 @@
+		adsLoader := NewAdsLoader()
+		svc := &ServiceEvent{
+			DynamicLoader: adsLoader,
+		}
+		cluster := &config_cluster_v3.Cluster{
+			Name: "ut-cluster",
+			ClusterDiscoveryType: &config_cluster_v3.Cluster_Type{
+				Type: config_cluster_v3.Cluster_EDS,
+			},
+		}
+		anyCluster, err := anypb.New(cluster)
+		assert.NoError(t, err)
+		rsp := &service_discovery_v3.DiscoveryResponse{
+			Resources: []*anypb.Any{
+				anyCluster,
+			},
+		}
+		err = svc.handleCdsResponse(rsp)
+		assert.NoError(t, err)
+		assert.Equal(t, []string{"ut-cluster"}, svc.DynamicLoader.edsClusterNames)
+		wantHash := hash.Sum64String(anyCluster.String())
+		actualHash := svc.DynamicLoader.ClusterCache.GetCdsHash(cluster.GetName())
+		assert.Equal(t, wantHash, actualHash)
+	})
+
+	t.Run("test2: new cluster, cluster type is not eds", func(t *testing.T) {
+		adsLoader := NewAdsLoader()
+		svc := &ServiceEvent{
+			DynamicLoader: adsLoader,
+		}
+		cluster := &config_cluster_v3.Cluster{
+			Name: "ut-cluster",
+			ClusterDiscoveryType: &config_cluster_v3.Cluster_Type{
+				Type: config_cluster_v3.Cluster_LOGICAL_DNS,
+			},
+		}
+		anyCluster, err := anypb.New(cluster)
+		assert.NoError(t, err)
+		rsp := &service_discovery_v3.DiscoveryResponse{
+			Resources: []*anypb.Any{
+				anyCluster,
+			},
+		}
+		err = svc.handleCdsResponse(rsp)
+		assert.NoError(t, err)
+		assert.Equal(t, []string{}, svc.DynamicLoader.edsClusterNames)
+		wantHash := hash.Sum64String(anyCluster.String())
+		actualHash := svc.DynamicLoader.ClusterCache.GetCdsHash(cluster.GetName())
+		assert.Equal(t, wantHash, actualHash)
+	})
+
+	t.Run("test3: nil cluster", func(t *testing.T) {
+		adsLoader := NewAdsLoader()
+		svc := &ServiceEvent{
+			DynamicLoader: adsLoader,
+		}
+		cluster := &config_cluster_v3.Cluster{}
+		anyCluster, err := anypb.New(cluster)
+		assert.NoError(t, err)
+		rsp := &service_discovery_v3.DiscoveryResponse{
+			Resources: []*anypb.Any{
+				anyCluster,
+			},
+		}
+		err = svc.handleCdsResponse(rsp)
+		assert.NoError(t, err)
+		assert.Equal(t, []string{}, svc.DynamicLoader.edsClusterNames)
+		wantHash := hash.Sum64String(anyCluster.String())
+		actualHash := svc.DynamicLoader.ClusterCache.GetCdsHash(cluster.GetName())
+		assert.Equal(t, wantHash, actualHash)
+	})
+
+	t.Run("test4: already have cluster, change cluster hash", func(t *testing.T) {
+		adsLoader := NewAdsLoader()
+		svc := &ServiceEvent{
+			DynamicLoader: adsLoader,
+		}
+		cluster := &config_cluster_v3.Cluster{
+			Name: "ut-cluster",
+			ClusterDiscoveryType: &config_cluster_v3.Cluster_Type{
+				Type: config_cluster_v3.Cluster_LOGICAL_DNS,
+			},
+		}
+		anyCluster, err := anypb.New(cluster)
+		assert.NoError(t, err)
+		clusterHash := hash.Sum64String(anyCluster.String())
+		svc.DynamicLoader.ClusterCache.SetCdsHash(cluster.GetName(), clusterHash)
+		apiStatus := core_v2.ApiStatus_UPDATE
+		svc.DynamicLoader.CreateApiClusterByCds(apiStatus, cluster)
+
+		cluster.Name = "cluster"
```

Comment:
I think this is not updating cluster, this is remove a old one, and then add a new one.

As multi clusters can be returned, the test case multiplexed